File: absolute-pitch/AbsolutePitch.java
import javax.sound.sampled.*;
import java.util.*;
import javax.swing.*;
import javax.swing.table.*;
import javax.swing.event.*;
import java.awt.*;
import java.awt.event.*;
import javax.swing.plaf.metal.MetalLookAndFeel;

public class AbsolutePitch extends JApplet {
	public static void main(String[] args) {
		SwingUtilities.invokeLater(new Runnable() {
			public void run() {
				JFrame frame = new JFrame("Absolute (?) pitch training");

				UIManager.put("swing.boldMetal", Boolean.FALSE);
				try {
					UIManager.setLookAndFeel(new MetalLookAndFeel());
				} catch (UnsupportedLookAndFeelException e) {
				}
				SwingUtilities.updateComponentTreeUI(frame);

				frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
				AbsolutePitchPanel app = new AbsolutePitchPanel();
				frame.add(app);
				frame.pack();
				frame.setVisible(true);
				app.start();
			}
		});
	}

	private AbsolutePitchPanel app;

	@Override
	public void init() {
		app = new AbsolutePitchPanel();
		getContentPane().add(app, BorderLayout.CENTER);
	}

	@Override
	public void start() {
		app.start();
	}

	@Override
	public void stop() {
		app.stop();
	}

	@Override
	public void destroy() {
		app.destroy();
	}

	public static class AbsolutePitchPanel extends JPanel {
		private SourceDataLine audioLine;
		private Random random = new Random();
		private JLabel answerLabel;

		private String[] pitchName = {
			"a", "a♯",  "c♭",  "c",   "c♯",  "d", "e♭",  "e",   "f",   "f♯",  "g", "a♭"
			// "a", "b",   "h",   "c",   "cis", "d", "es",  "e",   "f",   "fis", "g", "as"
			// "a", "ais", "ces", "his", "des", "d", "dis", "fes", "eis", "ges", "g", "gis"
		};
		private double frequencyPitch0 = 440;
		private int pitchMin = -16, pitchMax = 8;

		private int currentPitch;
		private boolean canAnswer = false;

		private int[][] count = new int[12][12];
		private MyTableModel tableModel;
		private JTable table;

		private double angle = 0;

		private static double RA(double f) {
			return 12200 * 12200 * f * f * f * f / (f * f + 20.6 * 20.6) / (f * f + 12200 * 12200) / Math.sqrt((f * f + 107.7 * 107.7) * (f * f + 737.9 * 737.9));
		}

		private byte sample(double dt, double p, double vol) {
			double frequency = frequencyPitch0 * Math.pow(2, (p / 12));
			angle += 2 * Math.PI * dt * frequency;
			vol *= Math.min(0.2 * 1.75 / (RA(frequency) + RA(frequency * 2) * 0.5 + RA(frequency * 4) * 0.25), 1);
			double s = Math.sin(angle); // + 0.1 * Math.sin(angle * 2) + 0.03 * Math.sin(angle * 4);
			s = (s < 0 ? -1 : 1) * Math.pow(s < 0 ? -s : s, 0.8);
			return (byte)(int)(127 * s * vol - random.nextInt(2));
		}

		private int randomPitch() {
			return pitchMin + random.nextInt(pitchMax - pitchMin);
		}

		private void playNext() {
			if (audioLine == null) {
				return;
			}
			new PlayerThread().start();
		}

		public void start() {
			if (audioLine != null) {
				audioLine.start();
				playNext();
			}
		}

		public void stop() {
			if (audioLine != null) {
				// audioLine.drain();
				audioLine.stop();
			}
		}

		public void destroy() {
			if (audioLine != null) {
				audioLine.close();
				audioLine = null;
			}
		}

		private class PlayerThread extends Thread {
			@Override
			public void run() {
				synchronized(audioLine) {
					int p0 = randomPitch();
					int rate = (int)audioLine.getFormat().getSampleRate();
					int n = 0;
					int distractionLength = rate * 9 / 2;
					byte[] data = new byte[distractionLength + 30 * (rate / 10)];
					for (int i = 0, len = rate / 10; i < len; ++i) {
						data[n++] = (sample(1.0 / rate, p0, (double)i / len));
					}
					for (int len, lenMax = distractionLength; (len = Math.min(lenMax, rate / (2 + random.nextInt(9)))) > 0; lenMax -= len) {
						if (len + rate / 10 > lenMax) {
							len = lenMax;
						}
						int p1 = randomPitch();
						for (int i = 0; i < len; ++i) {
							data[n++] = (sample(1.0 / rate, p0 + (p1 - p0) * (double)i / len, 1));
						}
						p0 = p1;
					}
					int m = n;
					for (int i = 0, len = 28 * (rate / 10); i < len; ++i) {
						data[n++] = (sample(1.0 / rate, p0, 1));
					}
					for (int i = 0, len = rate / 10; i < len; ++i) {
						data[n++] = (sample(1.0 / rate, p0, (1 - (double)i / len)));
					}
					currentPitch = ((p0 % 12) + 12) % 12;

					audioLine.start();
					audioLine.write(data, 0, m);
					audioLine.write(data, m, rate);
					answerLabel.setText("Click on the right note (table column)!");
					canAnswer = true;
					m += rate;
					audioLine.write(data, m, n - m);
				}
			}
		}

		private void answer(int pitch) {
			if (!canAnswer) {
				return;
			}
			pitch = ((pitch % 12) + 12) % 12;
			int error = ((pitch - currentPitch) % 12 + 12 + 6) % 12 - 6;
			count[currentPitch][(error + 12) % 12] += 1;
			int countRight = 0, countAll = 0;
			for (int[] t: count) {
				countRight += t[0];
				for (int i: t) {
					countAll += i;
				}
			}
			answerLabel.setText(
				"Accuracy: " + countRight + "/" + countAll + " = " + (int)(100.0 * countRight / countAll + 0.5) + "%. " +
				"Last: " + pitchName[currentPitch] + ". " +
				(
					error != 0
					? "Your guess (" + pitchName[pitch] + ") was " + error + " semitones too " + (error > 0 ? "high" : "low") + "."
					: "+1 point!"
				)
			);
			tableModel.change(currentPitch, error);
			canAnswer = false;
			playNext();
		}

		private class MyTableModel extends AbstractTableModel {
			@Override
			public int getColumnCount() {
				return 15;
			}

			@Override
			public int getRowCount() {
				return 13;
			}

			@Override
			public Class getColumnClass(int i) {
				return Integer.class;
			}

			@Override
			public String getColumnName(int i) {
				switch (i) {
					case 0: return "error";
					case 1: return "total";
					case 2: return "";
				}
				return pitchName[i - 3];
			}

			@Override
			public Object getValueAt(int row, int col) {
				int pitch = col - 3;
				int error = 5 - row;
				if (row == 12) {
					switch (col) {
						case 0: return "acc.";
						case 1: {
							int s = 0, all = 0;
							for (int[] t: count) {
								s += t[0];
								for (int i: t) {
									all += i;
								}
							}
							return (int)(100.0 * s / all + 0.5) + "%";
						}
						case 2: return "";
					}
					int s = 0;
					for (int i: count[pitch]) {
						s += i;
					}
					return (int)(100.0 * count[pitch][0] / s + 0.5) + "%";
				}
				switch (col) {
					case 0: return error > 0 ? ("+" + error) : ("" + error);
					case 1: {
						int s = 0;
						for (int[] t: count) {
							s += t[(error + 12) % 12];
						}
						return s;
					}
					case 2: return "";
				}
				return count[pitch][(error + 12) % 12];
			}

			public void change(int pitch, int error) {
				int row = 5 - error; // 5 - row == error
				int totalCol = 1;
				int pitchCol = pitch + 3;
				table.clearSelection();
				table.changeSelection(row, pitchCol, true, false);
				table.tableChanged(new TableModelEvent(this, row, row, totalCol, TableModelEvent.UPDATE));
				table.tableChanged(new TableModelEvent(this, row, row, pitchCol, TableModelEvent.UPDATE));
				row = 12;
				table.tableChanged(new TableModelEvent(this, row, row, totalCol, TableModelEvent.UPDATE));
				table.tableChanged(new TableModelEvent(this, row, row, pitchCol, TableModelEvent.UPDATE));
			}
		}

		public AbsolutePitchPanel() {
			setPreferredSize(new Dimension(640, 480));
			int sampleRate = 48000;
			AudioFormat format = new AudioFormat(sampleRate, 8, 1, true, false);
			try {
				audioLine = AudioSystem.getSourceDataLine(format);
				audioLine.open(format);
			} catch (LineUnavailableException e) {
				System.out.println(e);
				audioLine = null;
				this.add(new JLabel("Audio not available (" + sampleRate + " Hz 8-bit signed mono)"), BorderLayout.CENTER);
				return;
			}

			tableModel = new MyTableModel();
			table = new JTable(tableModel);

			MouseListener mouseListener = new MouseAdapter() {
				@Override
				public void mouseClicked(MouseEvent e) {
					int i = table.columnAtPoint(e.getPoint());
					if (3 <= i && i < 15) {
						answer(i - 3);
					}
				}
			};

			table.addMouseListener(mouseListener);
			table.getTableHeader().addMouseListener(mouseListener);
			table.getTableHeader().setResizingAllowed(false);
			table.getTableHeader().setReorderingAllowed(false);

			this.add(answerLabel = new JLabel("Listen to the sound and try to guess the pitch!"), BorderLayout.PAGE_START);
			this.add(new JScrollPane(table), BorderLayout.CENTER);
		}
	}
}

File: absolute-pitch/AbsolutePitch.py
#!/usr/bin/python3

import math
import random

notes = [
	"a", "b",   "h",   "c",   "cis", "d",  "es",   "e",   "f", "fis", "g", "as",
	"a", "ais", "ces", "his", "des", "d", "dis", "fes", "eis", "ges", "g", "gis",
	"a", "b",   "h",   "c",   "c#",  "d",  "eb",   "e",   "f", "f#",  "g", "ab",
	"a", "a#",  "cb",  "h#",  "db",  "d",  "d#",  "fb",  "e#", "gb",  "g", "g#"
]

import pyaudio

class AudioDevice:
	def __init__(self):
		self.pyaudio = pyaudio.PyAudio()
		self.rate = 48000
		self.stream = self.pyaudio.open(format=pyaudio.paInt8, channels=1, rate=self.rate, output=True)
	def play(self, data):
		self.stream.write(bytes(data))
	def close(self):
		self.stream.stop_stream()
		self.stream.close()
		self.pyaudio.terminate()

def generate_samples(p_min, p_max, rate):
	angle = 0
	def floatsample(dt, p):
		nonlocal angle
		frequency = 440 * (2 ** (p / 12))
		angle += 2 * math.pi * dt * frequency
		return math.sin(angle) + 0.03 * math.sin(angle * 2) + 0.01 * math.sin(angle * 4)

	def bytesample(f):
		return int(127 * f) & 0xff

	p0 = random.randint(p_min, p_max)
	data = bytearray()
	for i in range(0, rate // 10):
		data.append(bytesample(floatsample(1 / rate, p0) * i / (rate // 10)))
	for i in range(0, 20):
		p1 = random.randint(p_min, p_max)
		change_length = rate // random.randint(2, 10)
		for i in range(0, change_length):
			data.append(bytesample(floatsample(1 / rate, p0 + (p1 - p0) * i / change_length)))
		p0 = p1
	for i in range(0, rate * 2):
		data.append(bytesample(floatsample(1 / rate, p0)))
	for i in range(0, rate // 10):
		data.append(bytesample(floatsample(1 / rate, p0) * (1 - i / (rate // 10))))
	for i in range(0, 8 * rate // 10):
		data.append(bytesample(0))

	return p0, data

def stats_str(answers):
	s = [[]]
	s[-1].append("%4s" % "")
	for i in range(0, 12):
		s[-1].append("%4s" % notes[i])
	for j in range(5, -7, -1):
		s.append([])
		s[-1].append("%+4d" % j)
		for i in range(0, 12):
			if answers[i][j] > 0:
				s[-1].append("%4d" % answers[i][j])
			else:
				s[-1].append("%4s" % ".")
	s.append(s[0])
	return "\n".join("".join(row) for row in s)

def main():
	audio_device = AudioDevice()
	answers = dict((i, dict((j, 0) for j in range(-6, 6))) for i in range(0, 12))

	p_min = -16
	p_max = p_min - 1 + 12 * 2
	n = n_ok = 0
	try:
		while True:
			print(stats_str(answers))
			correct, samples = generate_samples(p_min, p_max, audio_device.rate)
			print("Listen closely, then enter the note name, or 'exit' to quit.")
			audio_device.play(samples)
			try:
				answer_str = input("? = ").lower()
				if answer_str in ("exit", "quit", "q"):
					break
				answer = notes.index(answer_str)
				offset = ((answer - correct + 6) % 12) - 6
				answers[correct % 12][offset] += 1
				if offset == 0:
					print("OK")
					n_ok += 1
				else:
					print("%+d off, correct = %s" % (offset, notes[correct % 12]))
			except ValueError:
				print("??, %s" % (notes[correct % 12]))
			n += 1
	except KeyboardInterrupt:
		pass
	except EOFError:
		pass
	finally:
		audio_device.close()

	print(stats_str(answers))
	if n > 0:
		print("n = %d, ok = %d, ratio = %.3f" % (n, n_ok, n_ok / n))
	else:
		print("n = 0")

if __name__ == "__main__":
	main()
